### DISCUSSION TITLE
Is it possible to use Calculated Value in GraphQL Filter?

### DISCUSSION THREAD

**Discussion by mjnewman** [pimcore/pimcore#7723] [Q&A / Support] [answered]
We're using a Calculated Value in an Object. The value is calculated from a Many-To-One Relation. We are checking to see if the object associated by the Many-To-One Relation is published.

The Calculated Value is working properly, we can see the correct value in the GraphQL Object Listing response. Using GraphQL we can SortBy the Calculated Value when the "Indexed" option is checked in the class definition.

When trying to use the Calculated Value as part of a GraphQL Filter it does not work, for example: **filter: "{\"ispublished\": \"yes\"}"** and **filter: "{\"ispublished\": {\"$not\": \"yes\"}}"** both return no results.

We are able to filter by other fields, such as Inputs and Text Areas.

Is it possible to use a Calculated Value as part of a GraphQL Filter? If so, are there any examples of how to go about doing it?

Any help would be greatly appreciated!

**Comment by weisswurstkanone** [comment]
Hi @mjnewman , can you provide a complete query example here please ? Thanks!

**Comment by mjnewman** [comment]
Hi @weisswurstkanone please find the query example below. I've removed some of the extraneous elements, but this should give a good example of what we're trying to do.

The Calculated Value **ispublished** is calculated from the **lookupobject->getPublished()** value and is populated with the string "true" or "false". With **ispublished** "Indexed" the SortBy/SortOrder is working. With "Indexed" unchecked SortBy/SortOrder does not appear to work.

The filter shown in the query returns no errors and a result of **"edges": []**. Without the filter, the query returns the expected results.

We're using DataHub version 0.6.3.

Thanks!

{
  getLookupListing(
    defaultLanguage: "en", 
      filter: "{\\"ispublished\\": \\"true\\"}", 
      sortBy: ["ispublished"], 
      sortOrder: ["asc"]
  ) {
    edges {
      node {
        ispublished
        title
        lookupobject {
          ... on object_Challenge {
            title
          }
        }
      }
    }
  }
}

**Reply by weisswurstkanone** [reply]
Works perfectly fine for me. Did you check what is in your object_query_[class id] table ?

**Reply by mjnewman** [reply]
The ispublished values in the object_query_[class id] table are NULL. The value in the GraphQL results appears as "true" or "false" as does the value in the object editor.

Perhaps I'm misunderstanding how to use this and how it works?

The code we're using is as follows (the same code is in getCalculatedValueForEditMode):

class CalculateValues implements CalculatorClassInterface {

  public function compute ( Concrete $object, CalculatedValue $context ):string {
    if ( $context->getFieldname() == 'ispublished' ) {
      $active = "false";
      $lookupobject = $object->getLookupobject();
      if ( $lookupobject && $lookupobject->getPublished() ) {
        $active = "true";
      }
      return $active;
    }
  }

**Comment by weisswurstkanone** [comment, marked as answer]
For the listings the query table is used. Although it says "calculated value" you have to at least save the object to populate the table.
If you just call the getter, then it will be calculated on the fly of course.

**Reply by mjnewman** [reply]
Thank you for explaining that and apologies for the confusion on my end. Your explanation makes perfect sense.

A bit of a tangent, is there a way to filter for a "null" value? We're looking to exclude GraphQL results when our "look up object" is not published – We're using scheduled events to publish these objects. When the object is not published, it returns null in the GraphQL results but we haven't found a way to filter for that.

For example, the following does not work:

    filter: "{\"lookupobject\": {\"$not\": null}}"

**Reply by weisswurstkanone** [reply]
Now it is, see
https://github.com/pimcore/data-hub/issues/299

**Reply by mjnewman** [reply]
Awesome – Thank you!

That works well when using it on the Computed Values, such as:

  filter: "{"ispublished": {"$not": null}}"

However it is not working on a Many-to-One relationship, such as:

  filter: "{"lookupobject": {"$not": null}}"

Do I have the filter set correctly or is this not possible with a Many-to-One relationship?

Sample from the GraphQL query:

edges {
  node {
    ispublished
    lookupobject {
      ... on object_MyObject {
        id
        title
      }
    }
  }
}

**Comment by weisswurstkanone** [comment]
Currently not supported as many-to-one occupies 2 columns in the query table.
This would require (probably a lot of) extra work.

**Reply by mjnewman** [reply]
Got it – Thank you for confirming. That makes sense.

Are there any other strategies for achieving that in GraphQL? I was hoping to solve it with the Calculated Value, which works, but not in my use case where the object is published as part of a scheduled event so the state is unknown when the parent object is first saved.

**Comment by weisswurstkanone** [comment]
Maybe the Pimcore\Bundle\DataHubBundle\Event\GraphQL\Model\ListEvent could be helpful which is fired before the list eventually gets loaded.

never used it though

https://github.com/pimcore/data-hub/commit/f49d1a39cabadb86d568aa714c9252450c168af5

**Reply by mjnewman** [reply]
I will check that out – Thanks again for your help!
